apm-sniffer/apm-sdk-plugin/rocketMQ-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v4/AbstractMessageConsumeInterceptor.java

```diff
@@ -48,7 +50,12 @@ public final void beforeMethod(EnhancedInstance objInst, Method method, Object[]
         ContextCarrier contextCarrier = getContextCarrierFromMessage(msgs.get(0));
         AbstractSpan span = ContextManager.createEntrySpan(CONSUMER_OPERATION_NAME_PREFIX + msgs.get(0)
                                                                                                 .getTopic() + "/Consumer", contextCarrier);
-
+        Tags.MQ_TOPIC.set(span, msgs.get(0).getTopic());
+        if (msgs.get(0).getStoreHost() != null) {
+            String brokerAddress = msgs.get(0).getStoreHost().toString();
+            brokerAddress = StringUtils.removeStart(brokerAddress, "/");
+            Tags.MQ_BROKER.set(span, brokerAddress);
```

Comment:
@pg-yang are you going to use this addr as virtual service? If so, I am concerned store host is not the complete broker list.
@zhyyu Could you share what is this like in a cluster mode deployment?

Reply:
> are you going to use this addr as virtual service

Yes, MQ_BROKER should be complete list

Reply:
I will try it later, today I am working over time.

Reply:
Take you time. It's not that urgent.

Reply:
I changed `configuration.yml` to implement cluster mode like below by followed the  [deployment doc](https://rocketmq.apache.org/docs/4.x/deployment/15deploy/) doc

```
type: jvm
entryService: http://localhost:8080/rocketmq-scenario/case/rocketmq-scenario
healthCheck: http://localhost:8080/rocketmq-scenario/case/healthCheck
startScript: ./bin/startup.sh
environment:
  - NAME_SERVER=namesrv:9876
depends_on:
  - namesrv
  - brokera
  - brokerb
dependencies:
  namesrv:
    image: apache/rocketmq:${CASE_SERVER_IMAGE_VERSION}
    hostname: namesrv
    command: sh mqnamesrv
  brokera:
    image: apache/rocketmq:${CASE_SERVER_IMAGE_VERSION}
    hostname: brokera
    command: sh mqbroker -n namesrv:9876 -c ../conf/2m-noslave/broker-a.properties
    depends_on:
      - namesrv
  brokerb:
    image: apache/rocketmq:${CASE_SERVER_IMAGE_VERSION}
    hostname: brokerb
    command: sh mqbroker -n namesrv:9876 -c ../conf/2m-noslave/broker-b.properties
    depends_on:
      - namesrv
```

The actualData.yaml like below

```
segmentItems:
- serviceName: rocketmq-scenario
  segmentSize: 3
  segments:
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.37.16671294232900000
    ...
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.38.16671294249190000
    spans:
    - operationName: RocketMQ/TopicTest/Producer
      ...
      tags:
      - {key: mq.broker, value: '172.18.0.3:10911'}
      - {key: mq.topic, value: TopicTest}
    - operationName: GET:/case/rocketmq-scenario
      parentSpanId: -1
      ...
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.87.16671294268870000
    spans:
    - operationName: RocketMQ/TopicTest/Consumer
      ...
      tags:
      - {key: transmission.latency, value: '1893'}
      - {key: mq.topic, value: TopicTest}
      - {key: mq.broker, value: '172.18.0.3:10911'}
      ...
```

The `mq.broker` tag value still a single node. That maybe rocketmq provider and consumer only concern nameserver list. Just like the rocketmq deployment doc say [deployment doc](https://rocketmq.apache.org/docs/4.x/deployment/15deploy/)

> 如上启动命令是在单个NameServer情况下使用的。对于多个NameServer的集群，Broker启动命令中-n后面的地址列表用分号隔开即可，例如 192.168.1.1:9876;192.161.2:9876。

Reply:
@zhyyu What is the `peer` value of RocketMQ for now? Is it a single host too? If so, I have concerned there is another issue in the topology.

Reply:
Could you hold the list(`192.168.1.1:9876;192.161.2:9876`) like JDBC did? Which caches this from the beginning through dynamic field, rather than using the API.
Right now, we should determine, whether broker address or peer is better. @zhyyu @pg-yang

Reply:
> whether broker address or peer is better.

broker address or peer  only  represents  one node , rather than a cluster . 

I  think it's OK to use `nameserver` list as brokers .

Reply:
The whole actualData.yaml like below, the provider has peer `namesrv:9876`, not consumer doesn't have a peer.

```
segmentItems:
- serviceName: rocketmq-scenario
  segmentSize: 3
  segments:
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.37.16671294232900000
    spans:
    - operationName: RocketMQ/HealthCheckTopicTest/Producer
      parentSpanId: 0
      spanId: 1
      spanLayer: MQ
      startTime: 1667129424630
      endTime: 1667129424881
      componentId: 38
      isError: false
      spanType: Exit
      peer: namesrv:9876
      skipAnalysis: false
      tags:
      - {key: mq.broker, value: '172.18.0.3:10911'}
      - {key: mq.topic, value: HealthCheckTopicTest}
    - operationName: HEAD:/case/healthCheck
      parentSpanId: -1
      spanId: 0
      spanLayer: Http
      startTime: 1667129423326
      endTime: 1667129424907
      componentId: 14
      isError: false
      spanType: Entry
      peer: ''
      skipAnalysis: false
      tags:
      - {key: url, value: 'http://localhost:8080/rocketmq-scenario/case/healthCheck'}
      - {key: http.method, value: HEAD}
      - {key: http.status_code, value: '200'}
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.38.16671294249190000
    spans:
    - operationName: RocketMQ/TopicTest/Producer
      parentSpanId: 0
      spanId: 1
      spanLayer: MQ
      startTime: 1667129425017
      endTime: 1667129425171
      componentId: 38
      isError: false
      spanType: Exit
      peer: namesrv:9876
      skipAnalysis: false
      tags:
      - {key: mq.broker, value: '172.18.0.3:10911'}
      - {key: mq.topic, value: TopicTest}
    - operationName: GET:/case/rocketmq-scenario
      parentSpanId: -1
      spanId: 0
      spanLayer: Http
      startTime: 1667129424919
      endTime: 1667129425189
      componentId: 14
      isError: false
      spanType: Entry
      peer: ''
      skipAnalysis: false
      tags:
      - {key: url, value: 'http://localhost:8080/rocketmq-scenario/case/rocketmq-scenario'}
      - {key: http.method, value: GET}
      - {key: http.status_code, value: '200'}
  - segmentId: c6514e62e4ae4c72b248e74d3abefb02.87.16671294268870000
    spans:
    - operationName: RocketMQ/TopicTest/Consumer
      parentSpanId: -1
      spanId: 0
      spanLayer: MQ
      startTime: 1667129426887
      endTime: 1667129426923
      componentId: 39
      isError: false
      spanType: Entry
      peer: ''
      skipAnalysis: false
      tags:
      - {key: transmission.latency, value: '1893'}
      - {key: mq.topic, value: TopicTest}
      - {key: mq.broker, value: '172.18.0.3:10911'}
      refs:
      - {parentEndpoint: 'GET:/case/rocketmq-scenario', networkAddress: 'namesrv:9876',
        refType: CrossProcess, parentSpanId: 1, parentTraceSegmentId: c6514e62e4ae4c72b248e74d3abefb02.38.16671294249190000,
        parentServiceInstance: 5c87291341ab409e84c09b7998b83f69@172.18.0.5, parentService: rocketmq-scenario,
        traceId: c6514e62e4ae4c72b248e74d3abefb02.38.16671294249190001}
meterItems:
- serviceName: rocketmq-scenario
  meterSize: 5
  meters:
  - meterId:
      name: thread_pool
      tags:
      - {name: metric_type, value: active_size}
      - {name: pool_name, value: tomcat_execute_pool}
    singleValue: 0.0
  - meterId:
      name: thread_pool
      tags:
      - {name: metric_type, value: queue_size}
      - {name: pool_name, value: tomcat_execute_pool}
    singleValue: 0.0
  - meterId:
      name: thread_pool
      tags:
      - {name: metric_type, value: pool_size}
      - {name: pool_name, value: tomcat_execute_pool}
    singleValue: 10.0
  - meterId:
      name: thread_pool
      tags:
      - {name: metric_type, value: core_pool_size}
      - {name: pool_name, value: tomcat_execute_pool}
    singleValue: 10.0
  - meterId:
      name: thread_pool
      tags:
      - {name: metric_type, value: max_pool_size}
      - {name: pool_name, value: tomcat_execute_pool}
    singleValue: 200.0
logItems: []

```

Reply:
Peer should be set, otherwise your topology breaks.
Check my posted issue and discussion.

Reply:
@zhyyu Please read from here, https://github.com/apache/skywalking/pull/9855#issuecomment-1297144347

And you could verify through the real OAP backend through its UI. Once you are running RocketMQ in the cluster mode, you should face multiple nodes(cubes) on the topology, one for a node.

Reply:
I have set the consumer span's peer to namesrv list like below` peer: namesrv1:9876;namesrv2:9876` . I tested it by using two nameserver nodes.

```
segmentItems:
- serviceName: rocketmq-scenario
  segmentSize: 4
  segments:
  ...
  - segmentId: 9c4b4bc597d74ceb8430d91138f02bb6.40.16673914085300000
    spans:
    - operationName: RocketMQ/TopicTest/Producer
      parentSpanId: 0
      spanId: 1
      spanLayer: MQ
      startTime: 1667391408586
      endTime: 1667391408710
      componentId: 38
      isError: false
      spanType: Exit
      peer: namesrv1:9876;namesrv2:9876
      skipAnalysis: false
      tags:
      - {key: mq.broker, value: '172.24.0.5:10911'}
      - {key: mq.topic, value: TopicTest}
    - operationName: GET:/case/rocketmq-scenario
      parentSpanId: -1
      spanId: 0
      spanLayer: Http
      ...
  - segmentId: 9c4b4bc597d74ceb8430d91138f02bb6.96.16673914107020000
    spans:
    - operationName: RocketMQ/TopicTest/Consumer
      parentSpanId: -1
      spanId: 0
      spanLayer: MQ
      startTime: 1667391410702
      endTime: 1667391410756
      componentId: 39
      isError: false
      spanType: Entry
      peer: namesrv1:9876;namesrv2:9876
      skipAnalysis: false
      tags:
      - {key: transmission.latency, value: '2155'}
      - {key: mq.topic, value: TopicTest}
      - {key: mq.broker, value: '172.24.0.5:10911'}
      refs:
      - {parentEndpoint: 'GET:/case/rocketmq-scenario', networkAddress: 'namesrv1:9876;namesrv2:9876',
        refType: CrossProcess, parentSpanId: 1, parentTraceSegmentId: 9c4b4bc597d74ceb8430d91138f02bb6.40.16673914085300000,
        parentServiceInstance: 098de6b125ab43cc83521584456e9b45@172.24.0.6, parentService: rocketmq-scenario,
        traceId: 9c4b4bc597d74ceb8430d91138f02bb6.40.16673914085300001}
```